// @po-ui/ng-components/esm2015/lib/components/po-field/po-clean/po-clean.component.js
import { __decorate, __metadata } from "tslib";
import { Component, ViewContainerRef } from '@angular/core';
import { getParentRef } from '../../../utils/util';
import { PoCleanBaseComponent } from './po-clean-base.component';
/**
 * @docsPrivate
 *
 * @docsExtends PoCleanBaseComponent
 *
 * @examplePrivate
 *
 * <example-private name="po-clean-labs" title="PO Clean Labs">
 *   <file name="sample-po-clean-labs.component.html"> </file>
 *   <file name="sample-po-clean-labs.component.ts"> </file>
 * </example-private>
 */
let PoCleanComponent = class PoCleanComponent extends PoCleanBaseComponent {
    constructor(viewRef) {
        super();
        this.viewRef = viewRef;
        this.parentComponent = getParentRef(this.viewRef);
    }
    setInputValue(value) {
        if (this.inputRef && this.inputRef.nativeElement) {
            this.inputRef.nativeElement.value = value;
        }
    }
    getInputValue() {
        if (this.inputRef && this.inputRef.nativeElement) {
            return this.inputRef.nativeElement.value;
        }
    }
};
PoCleanComponent.ctorParameters = () => [
    { type: ViewContainerRef }
];
PoCleanComponent = __decorate([
    Component({
        selector: 'po-clean',
        template: "<span *ngIf=\"showIcon()\" class=\"po-icon po-field-icon po-icon-close\" (click)=\"clear()\"> </span>\n"
    }),
    __metadata("design:paramtypes", [ViewContainerRef])
], PoCleanComponent);
export { PoCleanComponent };
