// @po-ui/ng-components/esm2015/lib/components/po-field/po-lookup/services/po-lookup-filter.service.js
import { __decorate, __metadata, __rest } from "tslib";
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { isTypeof } from '../../../../utils/util';
/**
 * @docsPrivate
 *
 * @description
 *
 * Serviço padrão utilizado para filtrar os dados do componente po-lookup.
 */
let PoLookupFilterService = class PoLookupFilterService {
    constructor(httpClient) {
        this.httpClient = httpClient;
        this.headers = new HttpHeaders({
            'X-PO-No-Message': 'true'
        });
    }
    getFilteredItems(filteredItemsParams) {
        const { filterParams } = filteredItemsParams, restFilteredItemsParams = __rest(filteredItemsParams, ["filterParams"]);
        const validatedFilterParams = this.validateParams(filterParams);
        const params = Object.assign(Object.assign({}, restFilteredItemsParams), validatedFilterParams);
        return this.httpClient.get(this.url, { headers: this.headers, params });
    }
    getObjectByValue(value, filterParams) {
        const validatedFilterParams = this.validateParams(filterParams);
        return this.httpClient.get(`${this.url}/${value}`, { headers: this.headers, params: validatedFilterParams });
    }
    setUrl(url) {
        this.url = url;
    }
    validateParams(params) {
        return isTypeof(params, 'object') && !Array.isArray(params) ? params : undefined;
    }
};
PoLookupFilterService.ctorParameters = () => [
    { type: HttpClient }
];
PoLookupFilterService = __decorate([
    Injectable(),
    __metadata("design:paramtypes", [HttpClient])
], PoLookupFilterService);
export { PoLookupFilterService };
