// @po-ui/ng-components/esm2015/lib/components/po-navbar/po-navbar-actions/po-navbar-action/po-navbar-action.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ViewContainerRef } from '@angular/core';
import { Router } from '@angular/router';
import { callFunction, getParentRef, isExternalLink, openExternalLink } from '../../../../utils/util';
let PoNavbarActionComponent = class PoNavbarActionComponent {
    constructor(viewContainerRef, router) {
        this.router = router;
        this.parentRef = getParentRef(viewContainerRef);
    }
    click() {
        if (this.action) {
            callFunction(this.action, this.parentRef || this);
            return;
        }
        if (this.link) {
            return this.openUrl(this.link);
        }
    }
    openUrl(url) {
        if (isExternalLink(url)) {
            return openExternalLink(url);
        }
        if (url) {
            return this.router.navigate([url]);
        }
    }
};
PoNavbarActionComponent.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: Router }
];
__decorate([
    Input('p-action'),
    __metadata("design:type", Function)
], PoNavbarActionComponent.prototype, "action", void 0);
__decorate([
    Input('p-icon'),
    __metadata("design:type", String)
], PoNavbarActionComponent.prototype, "icon", void 0);
__decorate([
    Input('p-label'),
    __metadata("design:type", String)
], PoNavbarActionComponent.prototype, "label", void 0);
__decorate([
    Input('p-link'),
    __metadata("design:type", String)
], PoNavbarActionComponent.prototype, "link", void 0);
__decorate([
    Input('p-tooltip-text'),
    __metadata("design:type", String)
], PoNavbarActionComponent.prototype, "tooltip", void 0);
PoNavbarActionComponent = __decorate([
    Component({
        selector: 'po-navbar-action',
        template: "<div tabindex=\"0\" class=\"po-navbar-action-content po-clickable\" (click)=\"click()\">\n  <span class=\"po-icon {{ icon }}\" [p-tooltip]=\"tooltip\"> </span>\n</div>\n"
    }),
    __metadata("design:paramtypes", [ViewContainerRef, Router])
], PoNavbarActionComponent);
export { PoNavbarActionComponent };
