// @po-ui/ng-components/esm5/lib/components/po-field/po-lookup/services/po-lookup-filter.service.js
import { __assign, __decorate, __metadata, __rest } from "tslib";
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { isTypeof } from '../../../../utils/util';
/**
 * @docsPrivate
 *
 * @description
 *
 * Serviço padrão utilizado para filtrar os dados do componente po-lookup.
 */
var PoLookupFilterService = /** @class */ (function () {
    function PoLookupFilterService(httpClient) {
        this.httpClient = httpClient;
        this.headers = new HttpHeaders({
            'X-PO-No-Message': 'true'
        });
    }
    PoLookupFilterService.prototype.getFilteredItems = function (filteredItemsParams) {
        var filterParams = filteredItemsParams.filterParams, restFilteredItemsParams = __rest(filteredItemsParams, ["filterParams"]);
        var validatedFilterParams = this.validateParams(filterParams);
        var params = __assign(__assign({}, restFilteredItemsParams), validatedFilterParams);
        return this.httpClient.get(this.url, { headers: this.headers, params: params });
    };
    PoLookupFilterService.prototype.getObjectByValue = function (value, filterParams) {
        var validatedFilterParams = this.validateParams(filterParams);
        return this.httpClient.get(this.url + "/" + value, { headers: this.headers, params: validatedFilterParams });
    };
    PoLookupFilterService.prototype.setUrl = function (url) {
        this.url = url;
    };
    PoLookupFilterService.prototype.validateParams = function (params) {
        return isTypeof(params, 'object') && !Array.isArray(params) ? params : undefined;
    };
    PoLookupFilterService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    PoLookupFilterService = __decorate([
        Injectable(),
        __metadata("design:paramtypes", [HttpClient])
    ], PoLookupFilterService);
    return PoLookupFilterService;
}());
export { PoLookupFilterService };
